src/components/TreeReactFlow/index.tsx

```diff
@@ -529,143 +637,446 @@ const makePrimerNode = async (
+      node.data.nodeData
+    );
+  }
+};
 
-// TreeReactFlow renders multiple definitions on one canvas.
-// For each definition, it displays three things:
-// - the definition's name
-// - the definition's type
-// - the definition's body (a term)
-// It ensures that these are clearly displayed as "one atomic thing",
-// i.e. to avoid confused readings that group the type of 'foo' with the body of 'bar' (etc)
-export const TreeReactFlow = (p: TreeReactFlowProps) => {
-  const [{ nodes, edges }, setLayout] = useState<
-    Graph<PrimerNodeWithDef, PrimerEdge>
-  >({
-    nodes: [],
-    edges: [],
-  });
+const makeSelectionFromNodeData = (
+  def: GlobalName,
+  id0: string,
+  nodeData: NodeData
+): Selection | undefined => {
+  const id = Number(id0);
+  // Non-numeric IDs correspond to non-selectable nodes (those with no ID in backend) e.g. pattern constructors.
+  if (!isNaN(id)) {
+    if (nodeData.tag == "termDefNode") {
+      return {
+        tag: "SelectionDef",
+        contents: {
+          def,
+          node: { meta: id, nodeType: nodeData.nodeType },
+        },
+      };
+    } else {
+      return {
+        tag: "SelectionTypeDef",
+        contents: {
+          def,
+          node: {
+            tag: "TypeDefConsNodeSelection",
+            contents: {
+              con: nodeData.con,
+              field: {
+                index: nodeData.index,
+                meta: id,
+              },
+            },
+          },
+        },
+      };
+    }
+  } else {
+    return undefined;
+  }
+};
 
-  useEffect(() => {
-    (async () => {
-      const [trees, nested] = await Promise.all(
-        p.defs.map<
-          Promise<
-            [
-              Tree<PrimerNodeWithDefNoPos, PrimerEdge>,
-              Graph<PrimerNodeWithDef, PrimerEdge>[]
-            ]
-          >
-        >(async (def) => {
-          const defNodeId = defNameToNodeId(def.name.baseName);
-          const sigEdgeId = "def-sig-" + def.name.baseName;
-          const bodyEdgeId = "def-body-" + def.name.baseName;
-          const defNameNode: PrimerNode = {
-            id: defNodeId,
+type PrimerNodeWithNested<N> = PrimerNode<
+  N & { nested: Graph<Positioned<PrimerNode<N>>, PrimerEdge>[] }
+>;
+type PrimerNodeWithNestedAndDef = PrimerNodeWithNested<{ def: GlobalName }>;
+
+const defToTree = async (
+  def: Def,
+  p: DefParams & NodeParams & { layout: LayoutParams }
+): Promise<Tree<PrimerNodeWithNestedAndDef, PrimerEdge>> => {
+  const defNodeId = defNameToNodeId(def.name.baseName);
+  const sigEdgeId = "def-sig-" + def.name.baseName;
+  const bodyEdgeId = "def-body-" + def.name.baseName;
+  const defNameNode: PrimerNodeWithNestedAndDef = {
+    id: defNodeId,
+    data: {
+      def: def.name,
+      width: p.nodeWidth * p.nameNodeMultipliers.width,
+      height: p.nodeHeight * p.nameNodeMultipliers.height,
+      selected: deepEqualTyped(p.selection, {
+        tag: "SelectionDef",
+        contents: { def: def.name },
+      }),
```

Comment:
We should actually be able to refactor this sort of thing so that we get `selected` from `makeSelectionFromNode`, which would would reduce a decent amount of duplication. We'd need to create just enough of the node, then wrap it with the `data.selected` value. I just don't know how to do this in a way that TypeScript is happy with. Something vaguely like:
```ts
const mkNode = (
  n: Exclude<PrimerNode, { data: { selected: boolean } }>,
  s: Selection
): PrimerNode => ({
  data: { selected: deepEqualTyped(makeSelectionFromNode(n), s) },
  ...n,
});
```